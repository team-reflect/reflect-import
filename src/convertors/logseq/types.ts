import {ConversionError} from '../../types'

export interface LogseqExport {
  version: number
  blocks: LogseqNote[]
}

export interface LogseqNote {
  'page-name': string
  id: string
  children: LogseqBlock[]
}

export type LogseqProperties = Record<string, string | number | Array<string>>

export interface LogseqBlock {
  format: string
  content: string
  properties: LogseqProperties | null
  id: string
  children: LogseqBlock[]
}

export class LogseqConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'LogseqConversionError'
  }
}
