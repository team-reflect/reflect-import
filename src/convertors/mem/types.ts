import {ConversionError} from '../../types'

export interface MemExportNote {
  id: string
  title: string
  markdown: string
  tags: string[]
  created: string
  updated: string
}

export type MemExport = MemExportNote[]

export class MemConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'MemConversionError'
  }
}
