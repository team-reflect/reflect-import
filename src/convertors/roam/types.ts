import {ConversionError, ConvertedNote} from '../../types'

export interface RoamNote {
  title: string
  uid: string
  'edit-time': number
  children?: RoamNoteString[]
}

export interface RoamNoteRef {
  uid: string
}
export interface RoamNoteString {
  string: string
  'create-time': number
  uid: string
  'edit-time': number
  refs?: RoamNoteRef[]
  children?: RoamNoteString[]
}

export interface RoamConvertedNote extends ConvertedNote {
  html: string
  // Make these properties required
  subject: string
  backlinkNoteIds: string[]
}

export class RoamConversionError extends ConversionError {
  constructor(message: string) {
    super(message)
    this.name = 'RoamConversionError'
  }
}
