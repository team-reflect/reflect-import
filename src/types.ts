export const REFLECT_HOSTNAME = 'reflect.app'

export interface Backlink {
  id: string
  label: string
}

export interface ConvertedNote {
  // A unique id for the note
  id: string
  // A required html string
  html: string
  // An optional subject
  subject?: string
  // A list of note ids that are references from inside this note
  backlinks?: Backlink[]
  // The date the note was created
  createdAt?: number
  // The date the note was updated
  updatedAt?: number
  // Whether or not the note is a daily note
  dailyAt?: number
}

export interface ConvertOptions {
  data: string
  filename?: string
  lastModified?: number
}

export interface ConvertedNoteError {
  id: string
  type: string
  message: string
}

export interface ConvertResponse {
  notes: ConvertedNote[]
  errors: ConvertedNoteError[]
}

export type ConvertorAcceptType = Record<string, string | string[]>

export interface Convertor {
  accept: ConvertorAcceptType
  convert(options: ConvertOptions): ConvertResponse
}
export class ConversionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ConversionError'
  }
}
