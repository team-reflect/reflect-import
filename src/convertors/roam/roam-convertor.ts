import {validateNotes} from 'helpers/validate'

import {RoamBacklinks} from './roam-backlinks'
import {RoamNoteConvertor} from './roam-note-convertor'
import {RoamConversionError, RoamConvertedNote, RoamNote} from './types'
import {Convertor} from '../../convertor'
import {ConvertOptions, ConvertResponse} from '../../types'

export class RoamConvertor extends Convertor {
  accept = {'application/json': ['.json']}
  description = 'Roam Research JSON'

  async convert({data}: ConvertOptions): Promise<ConvertResponse> {
    const roamNotes = JSON.parse(data) as RoamNote[]

    if (!Array.isArray(roamNotes)) {
      throw new RoamConversionError('Roam export must be an array of notes')
    }

    const backlinks = new RoamBacklinks(roamNotes)

    const notes = roamNotes.map((note) => this.convertRoamNote(note, backlinks))

    return validateNotes(notes)
  }

  private convertRoamNote(note: RoamNote, backlinks: RoamBacklinks): RoamConvertedNote {
    const convertor = new RoamNoteConvertor({
      graphId: this.graphId,
      linkHost: this.linkHost,
      backlinks,
      note,
    })

    return convertor.convert()
  }
}
