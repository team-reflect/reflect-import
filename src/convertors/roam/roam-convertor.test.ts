import {describe, it, expect} from 'vitest'

import {formatHtml} from '../../testing/format-html'
import dailyNotes from './fixtures/roam-daily-notes.json'
import exampleGraph from './fixtures/roam-example-graph.json'
import todoNotes from './fixtures/roam-todos.json'
import urlsNotes from './fixtures/roam-urls.json'
import {RoamConvertor} from './roam-convertor'
import {RoamNote} from './types'

describe('RoamConvertor', () => {
  const ROAM_SAMPLE: RoamNote = {
    title: 'The Great CEO Within',
    uid: '09-14-2020',
    'edit-time': 1600057101270,
    children: [
      {
        string: 'Author:: #[[Matt Mochary]]',
        'create-time': 1599161754087,
        uid: '-rC8kxJH4',
        'edit-time': 1600958531049,
        refs: [
          {
            uid: 'uuid123',
          },
        ],
      },
      {
        string: '**Chapter 1: Getting Started**',
        'create-time': 1576448157574,
        children: [
          {
            string:
              'This is explained clearly and thoroughly in Disciplined Entrepreneurship by Bill Aulet. I won’t repeat or even summarize what he wrote. If you haven’t yet launched or achieved more than $1 million of revenue, go read Bill’s book first.',
            'create-time': 1576093423770,
            uid: 'QhiYH-2V4',
            'edit-time': 1600716930617,
          },
        ],
        uid: 'bn1X-uy0x',
        'edit-time': 1600716930617,
      },
    ],
  }

  const htmlFromRoamNote = (note: RoamNote, graphId: string): string => {
    const notesJson = JSON.stringify([note])
    const convertor = new RoamConvertor({graphId})
    const {notes} = convertor.convert({data: notesJson})
    const [{html}] = notes
    return html
  }

  it('parseContentFromHTML', () => {
    const result = htmlFromRoamNote(ROAM_SAMPLE, '123')

    expect(formatHtml(result)).toMatchInlineSnapshot(
      `
      "<h1>The Great CEO Within</h1>
      <ul>
        <li>
          <p>
            Author::
            <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-uuid123\\"
              >Matt Mochary</a
            >
          </p>
        </li>
        <li>
          <p><strong>Chapter 1: Getting Started</strong></p>
          <ul>
            <li>
              <p>
                This is explained clearly and thoroughly in Disciplined
                Entrepreneurship by Bill Aulet. I won’t repeat or even summarize what
                he wrote. If you haven’t yet launched or achieved more than $1 million
                of revenue, go read Bill’s book first.
              </p>
            </li>
          </ul>
        </li>
      </ul>
      "
    `,
    )
  })

  describe('generateContentFromRoamNote', () => {
    it('handles both formats of todos', () => {
      const note: RoamNote = {
        title: 'note1',
        uid: '123',
        'edit-time': 12345,
        children: [
          {
            string: 'parent',
            'create-time': 12345,
            uid: '123',
            'edit-time': 12345,
            children: [
              {
                string: `{{TODO}}  todo 1`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
              {
                string: `{{[[TODO]]}}  todo 2`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
            ],
          },
        ],
      }

      const doc = htmlFromRoamNote(note, '123')
      expect(doc).toMatchSnapshot()
    })

    it('handles list with mixed bullets and tasks', () => {
      const note: RoamNote = {
        title: 'note1',
        uid: '123',
        'edit-time': 12345,
        children: [
          {
            string: 'parent',
            'create-time': 12345,
            uid: '123',
            'edit-time': 12345,
            children: [
              {
                string: `{{TODO}}  todo 1`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
              {
                string: `{{DONE}}  todo 2`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
              {
                string: `bullet 1`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
              {
                string: `bullet 2`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
              {
                string: `{{TODO}}  todo 3`,
                'create-time': 12345,
                uid: '456',
                'edit-time': 12345,
                children: [],
              },
            ],
          },
        ],
      }

      const doc = htmlFromRoamNote(note, '123')

      expect(doc).toMatchSnapshot()
    })
  })

  it('handles code blocks', () => {
    const node: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: '```javascript\nlonger code snippet\n```',
          'create-time': 12345,
          uid: '123',
          'edit-time': 12345,
        },
      ],
    }

    const html = htmlFromRoamNote(node, '123')

    expect(html).toMatchInlineSnapshot(`
    "<h1>note1</h1><ul><li><pre><code class=\\"language-javascript\\">longer code snippet
    </code></pre></li></ul>"
  `)
  })

  it('handles ordered lists', () => {
    const node: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: '1. foo',
          'create-time': 12345,
          uid: '123',
          'edit-time': 12345,
          children: [],
        },
      ],
    }

    const html = htmlFromRoamNote(node, '123')
    expect(html).toMatchInlineSnapshot(
      '"<h1>note1</h1><ul><li><p>1. foo</p><ul></ul></li></ul>"',
    )
  })

  it('handles blockquotes', () => {
    const node: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: '> hello world',
          'create-time': 12345,
          uid: '123',
          'edit-time': 12345,
          children: [],
        },
      ],
    }

    const html = htmlFromRoamNote(node, '123')
    expect(formatHtml(html)).toMatchInlineSnapshot(`
    "<h1>note1</h1>
    <ul>
      <li>
        <blockquote>
          <p>hello world</p>
        </blockquote>
        <ul></ul>
      </li>
    </ul>
    "
  `)
  })

  it('generates content with the roam note', () => {
    const note: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: 'hello world',
          'create-time': 12345,
          uid: '123',
          'edit-time': 12345,
          children: [
            {
              string: `everything is **awesome** #foo`,
              'create-time': 12345,
              uid: '456',
              'edit-time': 12345,
              children: [
                {
                  string: `wahooo!\n\ngood times`,
                  'create-time': 12345,
                  uid: '789',
                  'edit-time': 12345,
                },
              ],
            },
          ],
        },
      ],
    }

    const html = htmlFromRoamNote(note, '123')

    expect(formatHtml(html)).toMatchInlineSnapshot(`
    "<h1>note1</h1>
    <ul>
      <li>
        <p>hello world</p>
        <ul>
          <li>
            <p>
              everything is <strong>awesome</strong>
              <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/foo\\">foo</a>
            </p>
            <ul>
              <li>
                <p>wahooo!</p>
                <p>good times</p>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
    "
  `)
  })

  it('handles backlinks', () => {
    const note: RoamNote = {
      title: 'note1',
      uid: '123',
      'edit-time': 12345,
      children: [
        {
          string: 'link to [[another note]]',
          'create-time': 12345,
          uid: '123',
          'edit-time': 12345,
          children: [],
        },
      ],
    }

    const html = htmlFromRoamNote(note, '123')

    expect(html).toMatchInlineSnapshot(
      '"<h1>note1</h1><ul><li><p>link to <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/another note\\">another note</a></p><ul></ul></li></ul>"',
    )
  })

  it('parses todo', () => {
    const roamNote = todoNotes[0]
    const html = htmlFromRoamNote(roamNote, '123')

    expect(html).toMatchInlineSnapshot(
      '"<h1>test1</h1><ul><li><input type=\\"checkbox\\" ><p>item 1</p></li><li><input type=\\"checkbox\\" checked><p>item 2</p></li><li></li></ul>"',
    )
  })

  it('parses urls', () => {
    const roamNote = urlsNotes[0]
    const html = htmlFromRoamNote(roamNote, '123')

    expect(html).toMatchInlineSnapshot(
      '"<h1>urls</h1><ul><li><p><a href=\\"https://reflect.app\\">https://reflect.app</a></p></li><li></li></ul>"',
    )
  })

  it('parses daily notes', () => {
    const convertor = new RoamConvertor({graphId: '123'})
    const {notes} = convertor.convert({data: JSON.stringify(dailyNotes)})
    const [note] = notes

    expect(note.subject).toEqual('September 14th, 2020')
    expect(note.dailyAt).toEqual(1600041600000)
  })

  describe('generateListFromRoamNoteString', () => {
    it('doesnt fail with empty string', () => {
      const note: RoamNote = {
        title: 'note1',
        uid: '123',
        'edit-time': 12345,
        children: [
          {
            string: '',
            'create-time': 12345,
            uid: '123',
            'edit-time': 12345,
          },
        ],
      }

      const html = htmlFromRoamNote(note, '123')

      expect(html).toMatchInlineSnapshot('"<h1>note1</h1><ul><li></li></ul>"')
    })
  })

  describe('createdAt', () => {
    it('should be extracted from the title', () => {
      const note: RoamNote = {
        title: 'October 10th, 2020',
        uid: '123',
        'edit-time': 12345,
        children: [
          {
            string: '',
            'create-time': 12345,
            uid: '123',
            'edit-time': 12345,
          },
        ],
      }

      const convertor = new RoamConvertor({graphId: '123'})
      const {notes} = convertor.convert({data: JSON.stringify([note])})

      const [{createdAt, dailyAt}] = notes

      expect(createdAt).toEqual(new Date('2020-10-10T00:00:00.000Z').getTime())
      expect(dailyAt).toEqual(new Date('2020-10-10T00:00:00.000Z').getTime())
    })
  })

  it('parses exampleGraph', () => {
    const convertor = new RoamConvertor({graphId: '123'})
    const {notes} = convertor.convert({data: JSON.stringify(exampleGraph)})

    expect(notes).toMatchSnapshot()
  })
})
