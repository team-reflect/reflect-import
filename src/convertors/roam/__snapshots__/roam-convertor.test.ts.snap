// Vitest Snapshot v1

exports[`RoamConvertor > generateContentFromRoamNote > handles both formats of todos 1`] = `"<h1>note1</h1><ul><li><p>parent</p><ul><li><input type=\\"checkbox\\" ><p>todo 1</p><ul></ul></li><li><input type=\\"checkbox\\" ><p>todo 2</p><ul></ul></li></ul></li></ul>"`;

exports[`RoamConvertor > generateContentFromRoamNote > handles list with mixed bullets and tasks 1`] = `"<h1>note1</h1><ul><li><p>parent</p><ul><li><input type=\\"checkbox\\" ><p>todo 1</p><ul></ul></li><li><input type=\\"checkbox\\" checked><p>todo 2</p><ul></ul></li><li><p>bullet 1</p><ul></ul></li><li><p>bullet 2</p><ul></ul></li><li><input type=\\"checkbox\\" ><p>todo 3</p><ul></ul></li></ul></li></ul>"`;

exports[`RoamConvertor > parses exampleGraph 1`] = `
[
  {
    "backlinkNoteIds": [],
    "createdAt": 1600041600000,
    "dailyAt": 1600041600000,
    "html": "<h1>September 14th, 2020</h1><ul><li><p>Test</p></li><li></li></ul>",
    "id": "roam-09-14-2020",
    "subject": "September 14th, 2020",
    "updatedAt": 1600057101270,
  },
  {
    "backlinkNoteIds": [
      "roam-j6wBG1tYR",
    ],
    "createdAt": 1600128000000,
    "dailyAt": 1600128000000,
    "html": "<h1>September 15th, 2020</h1><ul><li><p>test</p></li><li><p>asdf <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-j6wBG1tYR\\">hid</a></p></li><li></li></ul>",
    "id": "roam-09-15-2020",
    "subject": "September 15th, 2020",
    "updatedAt": 1600139707013,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>hid</h1><ul></ul>",
    "id": "roam-j6wBG1tYR",
    "subject": "hid",
    "updatedAt": 1600150462125,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1600214400000,
    "dailyAt": 1600214400000,
    "html": "<h1>September 16th, 2020</h1><ul><li><p>New page:</p><ul><li><p>Hi there</p></li></ul></li><li><p>asdf</p></li><li><p>((fVSnYTWX-))</p></li><li><ul><li><p>asdfs</p></li></ul></li></ul>",
    "id": "roam-09-16-2020",
    "subject": "September 16th, 2020",
    "updatedAt": 1600171206515,
  },
  {
    "backlinkNoteIds": [
      "roam-20vV-tWxq",
      "roam-JSdjtHezQ",
    ],
    "createdAt": 1600560000000,
    "dailyAt": 1600560000000,
    "html": "<h1>September 20th, 2020</h1><ul><li><p>Blah</p></li><li><p>Antoher</p><ul><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-20vV-tWxq\\">Page</a></p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-JSdjtHezQ\\">Another one</a></p></li><li></li></ul></li></ul>",
    "id": "roam-09-20-2020",
    "subject": "September 20th, 2020",
    "updatedAt": 1600566771305,
  },
  {
    "backlinkNoteIds": [
      "roam-20vV-tWxq",
    ],
    "createdAt": 1600566888904,
    "dailyAt": undefined,
    "html": "<h1>Page</h1><ul><li><p>A comment</p><ul><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-20vV-tWxq\\">Page</a></p></li></ul></li></ul>",
    "id": "roam-20vV-tWxq",
    "subject": "Page",
    "updatedAt": 1600566782912,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1600566843402,
    "dailyAt": undefined,
    "html": "<h1>Another one</h1><ul><li></li></ul>",
    "id": "roam-JSdjtHezQ",
    "subject": "Another one",
    "updatedAt": 1600566787181,
  },
  {
    "backlinkNoteIds": [
      "roam-20vV-tWxq",
    ],
    "createdAt": 1601510400000,
    "dailyAt": 1601510400000,
    "html": "<h1>October 1st, 2020</h1><ul><li><p>A parent</p><ul><li><p>Alex <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-20vV-tWxq\\">Page</a></p></li></ul></li><li><p>Alex Page</p><ul><li></li><li><p>Test</p><ul><li><p>Another</p><ul><li><p>Deep</p><ul><li><p>asdfasdf</p></li><li><p>as</p></li><li><p>dfa</p></li><li><p>sdf'asd</p></li></ul></li></ul></li></ul></li></ul></li></ul>",
    "id": "roam-10-01-2020",
    "subject": "October 1st, 2020",
    "updatedAt": 1601547790265,
  },
  {
    "backlinkNoteIds": [
      "roam-20vV-tWxq",
    ],
    "createdAt": 1601596800000,
    "dailyAt": 1601596800000,
    "html": "<h1>October 2nd, 2020</h1><ul><li><p>asdf</p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-20vV-tWxq\\">Page</a></p></li><li><p>asdf</p><ul><li><p>asdf</p><ul><li><p>adsf</p></li></ul></li></ul></li></ul>",
    "id": "roam-10-02-2020",
    "subject": "October 2nd, 2020",
    "updatedAt": 1601603710972,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1601683200000,
    "dailyAt": 1601683200000,
    "html": "<h1>October 3rd, 2020</h1><ul></ul>",
    "id": "roam-10-03-2020",
    "subject": "October 3rd, 2020",
    "updatedAt": 1601636405186,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1601856000000,
    "dailyAt": 1601856000000,
    "html": "<h1>October 5th, 2020</h1><ul><li></li></ul>",
    "id": "roam-10-05-2020",
    "subject": "October 5th, 2020",
    "updatedAt": 1601865293407,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1602028800000,
    "dailyAt": 1602028800000,
    "html": "<h1>October 7th, 2020</h1><ul></ul>",
    "id": "roam-10-07-2020",
    "subject": "October 7th, 2020",
    "updatedAt": 1602030149795,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1602547200000,
    "dailyAt": 1602547200000,
    "html": "<h1>October 13th, 2020</h1><ul><li><p>It's a daily note</p><ul><li><p>asdfasdf <strong>bold</strong> sdf</p></li><li></li></ul></li></ul>",
    "id": "roam-10-13-2020",
    "subject": "October 13th, 2020",
    "updatedAt": 1602564338609,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1602633600000,
    "dailyAt": 1602633600000,
    "html": "<h1>October 14th, 2020</h1><ul></ul>",
    "id": "roam-10-14-2020",
    "subject": "October 14th, 2020",
    "updatedAt": 1602598553692,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1602720000000,
    "dailyAt": 1602720000000,
    "html": "<h1>October 15th, 2020</h1><ul></ul>",
    "id": "roam-10-15-2020",
    "subject": "October 15th, 2020",
    "updatedAt": 1602673203764,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1605312000000,
    "dailyAt": 1605312000000,
    "html": "<h1>November 14th, 2020</h1><ul><li></li></ul>",
    "id": "roam-11-14-2020",
    "subject": "November 14th, 2020",
    "updatedAt": 1605310990419,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1606089600000,
    "dailyAt": 1606089600000,
    "html": "<h1>November 23rd, 2020</h1><ul><li><p>test</p></li><li><p>asdf</p><ul><li><p>asdf</p></li><li><p>asdf</p></li><li></li><li><p>asdfasdf</p></li><li><ul><li><p>dasdfasdf</p><ul><li><p>asdf</p></li></ul></li></ul></li></ul></li><li><p>asdf</p></li><li></li></ul>",
    "id": "roam-11-23-2020",
    "subject": "November 23rd, 2020",
    "updatedAt": 1606096906114,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1607385600000,
    "dailyAt": 1607385600000,
    "html": "<h1>December 8th, 2020</h1><ul></ul>",
    "id": "roam-12-08-2020",
    "subject": "December 8th, 2020",
    "updatedAt": 1607401158714,
  },
  {
    "backlinkNoteIds": [
      "roam-JSdjtHezQ",
    ],
    "createdAt": 1607817600000,
    "dailyAt": 1607817600000,
    "html": "<h1>December 13th, 2020</h1><ul><li><p>one</p></li><li><p>two</p></li><li><p>tthree</p></li><li><p>asd</p><ul><li><p>d</p></li></ul></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-JSdjtHezQ\\">Another one</a></p></li><li></li><li></li><li></li><li></li><li></li></ul>",
    "id": "roam-12-13-2020",
    "subject": "December 13th, 2020",
    "updatedAt": 1607800050272,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1607800116658,
    "dailyAt": undefined,
    "html": "<h1>Roam</h1><ul><li></li></ul>",
    "id": "roam-C6BuI5xV7",
    "subject": "Roam",
    "updatedAt": 1607800090084,
  },
  {
    "backlinkNoteIds": [
      "roam-pTKChrTZg",
    ],
    "createdAt": 1612569600000,
    "dailyAt": 1612569600000,
    "html": "<h1>February 6th, 2021</h1><ul><li><p>Hello <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-pTKChrTZg\\">Richard</a></p></li><li><p>I can't <em>even</em> do html <strong>preview</strong></p><ul><li></li></ul></li></ul>",
    "id": "roam-02-06-2021",
    "subject": "February 6th, 2021",
    "updatedAt": 1612558450851,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>Richard</h1><ul></ul>",
    "id": "roam-pTKChrTZg",
    "subject": "Richard",
    "updatedAt": 1612558460650,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1612828800000,
    "dailyAt": 1612828800000,
    "html": "<h1>February 9th, 2021</h1><ul></ul>",
    "id": "roam-02-09-2021",
    "subject": "February 9th, 2021",
    "updatedAt": 1612811843299,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1616112000000,
    "dailyAt": 1616112000000,
    "html": "<h1>March 19th, 2021</h1><ul><li><p>one</p><ul><li><p>asdf</p></li><li><p>asdf</p><ul><li></li></ul></li></ul></li></ul>",
    "id": "roam-03-19-2021",
    "subject": "March 19th, 2021",
    "updatedAt": 1616138582207,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1618185600000,
    "dailyAt": 1618185600000,
    "html": "<h1>April 12th, 2021</h1><ul><li><p>asdf</p><ul><li><p>asdf</p><ul><li></li></ul></li></ul></li></ul>",
    "id": "roam-04-12-2021",
    "subject": "April 12th, 2021",
    "updatedAt": 1618185082951,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1622937600000,
    "dailyAt": 1622937600000,
    "html": "<h1>June 6th, 2021</h1><ul></ul>",
    "id": "roam-06-06-2021",
    "subject": "June 6th, 2021",
    "updatedAt": 1623029435620,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1624406400000,
    "dailyAt": 1624406400000,
    "html": "<h1>June 23rd, 2021</h1><ul><li><p>one</p><ul><li><p>two</p></li><li><p><em>three</em></p></li><li></li></ul></li></ul>",
    "id": "roam-06-23-2021",
    "subject": "June 23rd, 2021",
    "updatedAt": 1624463476606,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1627257600000,
    "dailyAt": 1627257600000,
    "html": "<h1>July 26th, 2021</h1><ul><li><p>Daily note</p></li></ul>",
    "id": "roam-07-26-2021",
    "subject": "July 26th, 2021",
    "updatedAt": 1627310438148,
  },
]
`;

exports[`RoamConvertor > parses exampleGraph2 1`] = `
[
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>Anonymous</h1><ul></ul>",
    "id": "roam-kfEoMevXn",
    "subject": "Anonymous",
    "updatedAt": 1669965625215,
  },
  {
    "backlinkNoteIds": [
      "roam-DOBbkj36v",
      "roam-KdxCBc-jS",
    ],
    "createdAt": 1669939200000,
    "dailyAt": 1669939200000,
    "html": "<h1>December 2nd, 2022</h1><ul><li><p>First page</p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-DOBbkj36v\\">Richard</a> is awesome</p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-KdxCBc-jS\\">Person</a> connected to ((OekWfoURN)) mentioning <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-DOBbkj36v\\">Richard</a></p></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-DOBbkj36v\\">Richard</a> ((LcbA2Kdnp)) and is a <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-KdxCBc-jS\\">Person</a></p></li><li><p>((LcbA2Kdnp)) brother <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-DOBbkj36v\\">Richard</a> is a <a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/roam-KdxCBc-jS\\">Person</a></p></li><li><input type=\\"checkbox\\" checked><p>asdfdd</p></li><li><input type=\\"checkbox\\" ><p>todo</p></li></ul>",
    "id": "roam-12-02-2022",
    "subject": "December 2nd, 2022",
    "updatedAt": 1669965625219,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1669965645817,
    "dailyAt": undefined,
    "html": "<h1>Richard</h1><ul><li><p>test</p></li><li><p>Is my brother</p></li></ul>",
    "id": "roam-DOBbkj36v",
    "subject": "Richard",
    "updatedAt": 1669965639615,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": 1669965651533,
    "dailyAt": undefined,
    "html": "<h1>Person</h1><ul><li><p>Something excellent about people</p></li><li><p>A person is a person</p></li></ul>",
    "id": "roam-KdxCBc-jS",
    "subject": "Person",
    "updatedAt": 1669965644221,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>ric</h1><ul></ul>",
    "id": "roam-MH_5XfEZ-",
    "subject": "ric",
    "updatedAt": 1669965841744,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>TODO</h1><ul></ul>",
    "id": "roam-ijCI34X3F",
    "subject": "TODO",
    "updatedAt": 1669973027789,
  },
  {
    "backlinkNoteIds": [],
    "createdAt": undefined,
    "dailyAt": undefined,
    "html": "<h1>DONE</h1><ul></ul>",
    "id": "roam-U4DhTc5Fh",
    "subject": "DONE",
    "updatedAt": 1669973038813,
  },
]
`;
